Issue #4550 XmlConfiguration argument matching

gregw opened the pull request:
Improve #4550 argument matching by:
 + rejecting obviously non matches (with allowance for unboxing)
 + sorting methods so that derived arguments are tried before more generic (eg String before Object)

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4550 XmlConfiguration argument matching

Improve argument matching by:
 + rejecting obviously non matches (with allowance for unboxing)
 + sorting methods so that derived arguments are tried before more generic (eg String before Object)

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
This change is to address the issue noticed by @joakime that:
>  the <Call name="add"> was using public boolean AbstractList.add(E e); not the one in Configurations.add(String ... configClassNames)

I think this is kind of a bad class design, but now at least we will deterministically favour a `String` over `Object` over `String...`.  Note this doesn't fix the problem @joakime found.  We would have to favour `String...` over `Object` to do that.   Hmmm maybe we should ?

gregw commented:
I have the order correct as the following test:
```java
    private static class TestOrder
    {
        public void other(Object o)
        {
            System.err.println("object");
        }

        public void other(String... s)
        {
            System.err.println("string varargs");
        }
    }

    @Test
    public void testMethodOrdering() throws Exception
    {
        TestOrder to = new TestOrder();
        to.other("foo");
    }
```

prints `object`

gregw commented:
Dang! still breaks things... stand by...

gregw committed:
Issue #4550 XmlConfiguration argument matching

Improve argument matching by:
 + can unbox from any Number to any Number

Signed-off-by: Greg Wilkins <gregw@webtide.com>

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/TypeUtil.java:
```diff
@@ -60,6 +60,22 @@
+        unbox.put(boolean.class, Boolean.class);
+        __unbox = Collections.unmodifiableMap(unbox);
+    }
+
```
`short` maps to `Long`! The list should be sorted by smaller to bigger types (i.e. from `boolean` to `double`) so it's easier to see that they are all there.

Merge this `static` block with the one below.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/TypeUtil.java:
```diff
@@ -727,4 +743,16 @@ public static URI getModuleLocation(Class<?> clazz)
+
+        Class<?> c = __unbox.get(type);
+        Class<?> ac = arg.getClass();
+        return ac == __unbox.get(type) || (Number.class.isAssignableFrom(c) && Number.class.isAssignableFrom(ac));
```
Not sure I get this? 

First expression should be just `ac == c`.

For `c==Byte.class` and `ac==Long.class` this returns `true`, but it is what we really want?
We cannot unbox a `long` into a `byte`, right?
Also there are other types implementing `Number` such as `Atomic*` and `Big[Integer|Decimal]`.
Calling `isUnboxable(int.class, null)` returns true, but we cannot unbox `null` to an `int`.
Calling `isUnboxed(int.class, new BigDecimal("0"))` returns true.

This feels it has a very specific logic that is only used in `XmlConfiguration` so perhaps should be there rather than in this class?

sbordet commented on jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java:
```diff
@@ -1734,47 +1756,64 @@ Args asEmptyVarArgs(Class<?> varArgType)
+                        !params[i].getType().isAssignableFrom(args[i].getClass()) &&
+                        !TypeUtil.isUnboxable(params[i].getType(), args[i]))
+                        return null;
+                }
```
I'm not sure I understand this.
We want to return `args` if unboxing is possible? That is already done by reflection.
Removing this block of code does not fail any test, so I am not sure why it was added?

gregw committed:
Issue #4550

Do not check the assignability of the arguments.  Instead rely on the order of the methods.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4550

unbox test no longer required

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4550

Simplified test

Signed-off-by: Greg Wilkins <gregw@webtide.com>

sbordet commented on jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java:
```diff
@@ -99,20 +100,42 @@
         }
+
         return compare;
     };
```
How about this:

```java
    public static final Comparator<Executable> EXECUTABLE_COMPARATOR = (e1, e2) ->
    {
        int count = e1.getParameterCount();
        int compare = Integer.compare(count, e2.getParameterCount());
        if (compare == 0 && count > 0)
        {
            Parameter[] p1 = e1.getParameters();
            Parameter[] p2 = e2.getParameters();
            boolean va1 = p1[count - 1].isVarArgs();
            boolean va2 = p2[count - 1].isVarArgs();
            compare = Boolean.compare(va1, va2);
            if (compare == 0)
            {
                // Rank by assignability of the arguments
                for (int i = 0; i < count; i++)
                {
                    Class<?> t1 = p1[i].getType();
                    Class<?> t2 = p2[i].getType();
                    if (t1 != t2)
                    {
                        int weight = 2 ^ (count - 1 - i);
                        boolean ref1 = !t1.isPrimitive();
                        boolean ref2 = !t2.isPrimitive();
                        int refs = Boolean.compare(ref1, ref2);
                        if (refs == 0)
                        {
                            if (t1.isAssignableFrom(t2))
                                compare += weight;
                            else if (t2.isAssignableFrom(t1))
                                compare -= weight;
                        }
                        else
                        {
                            compare += refs * weight;
                        }
                    }
                }
            }
            compare = Math.min(1, Math.max(compare, -1));
        }
        return compare;
    };
```

gregw committed:
Issue #4550

Cleanup comparator

Signed-off-by: Greg Wilkins <gregw@webtide.com>

sbordet commented on jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java:
```diff
@@ -99,20 +100,47 @@
+                        if (compare == 0)
+                        {
+                            // favour primitive type over reference
+                            compare = Boolean.compare(t2.isPrimitive(), t1.isPrimitive());
```
It's subtle that `t1` and `t2` are inverted here.
I would prefer `Boolean.compare(!t1.isPrimitive(), !t2.isPrimitive())`.

gregw replied:
that's how I wrote it initially.... switching....

sbordet reviewed (approved):
Other than a nit in the comparator, LGTM.

gregw committed:
Issue #4550

Cleanup comparator

Signed-off-by: Greg Wilkins <gregw@webtide.com>
